src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java

```diff
@@ -113,177 +139,313 @@ protected AbstractApiAction(
+        throws IOException {
+        return loadConfiguration(getConfigType(), false, logComplianceEvent).map(
+            configuration -> ValidationResult.success(SecurityConfiguration.of(entityName, configuration))
+        );
+    }
 
-        if (isReadonlyFieldUpdated(existingConfiguration, content)) {
-            conflict(channel, "Attempted to update read-only property.");
-            return;
+    protected final ValidationResult<SecurityDynamicConfiguration<?>> loadConfiguration(
+        final CType cType,
+        boolean omitSensitiveData,
+        final boolean logComplianceEvent
+    ) {
+        final var configuration = load(cType, logComplianceEvent);
+        if (configuration.getSeqNo() < 0) {
+            return ValidationResult.error(
+                RestStatus.FORBIDDEN,
+                forbiddenMessage(
+                    "Security index need to be updated to support '" + getConfigType().toLCString() + "'. Use SecurityAdmin to populate."
+                )
+            );
         }
-
-        if (LOGGER.isTraceEnabled() && content != null) {
-            LOGGER.trace(content.toString());
+        if (omitSensitiveData) {
+            if (!isSuperAdmin()) {
+                configuration.removeHidden();
+            }
+            configuration.clearHashes();
+            configuration.set_meta(null);
         }
+        return ValidationResult.success(configuration);
+    }
 
-        boolean existed = existingConfiguration.exists(name);
-        final Object newContent = DefaultObjectMapper.readTree(content, existingConfiguration.getImplementingClass());
-        if (!hasPermissionsToCreate(existingConfiguration, newContent, getResourceName())) {
-            forbidden(channel, "No permissions");
-            return;
+    protected final ValidationResult<Pair<User, TransportAddress>> withUserAndRemoteAddress() {
+        final var userAndRemoteAddress = Utils.userAndRemoteAddressFrom(threadPool.getThreadContext());
+        if (userAndRemoteAddress.getLeft() == null) {
+            return ValidationResult.error(RestStatus.UNAUTHORIZED, payload(RestStatus.UNAUTHORIZED, "Unauthorized"));
         }
-        existingConfiguration.putCObject(name, newContent);
-
-        AbstractApiAction.saveAndUpdateConfigs(
-            this.securityIndexName,
-            client,
-            getConfigName(),
-            existingConfiguration,
-            new OnSucessActionListener<IndexResponse>(channel) {
-
-                @Override
-                public void onResponse(IndexResponse response) {
-                    if (existed) {
-                        successResponse(channel, "'" + name + "' updated.");
-                    } else {
-                        createdResponse(channel, "'" + name + "' created.");
-                    }
+        return ValidationResult.success(userAndRemoteAddress);
+    }
 
-                }
+    protected EndpointValidator createEndpointValidator() {
```

Comment:
Is this validator unused, seems like its has no code coverage?

Reply:
Yes due to the problem with mocking this class, but your idea with `AbstractApiActionDependencies` class "open gates" to do it :-)